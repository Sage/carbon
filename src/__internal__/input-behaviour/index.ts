export { InputBehaviour, InputContext } from "./input-behaviour.component";
export {
  InputGroupBehaviour,
  InputGroupContext,
} from "./input-group-behaviour.component";
export type { InputContextProps } from "./useInputBehaviour";
