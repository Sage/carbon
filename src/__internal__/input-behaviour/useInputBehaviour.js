import { useState, useRef, useCallback, useMemo } from "react";

const useInputBehaviour = (blockGroupBehaviour) => {
  const [hasFocus, setHasFocus] = useState(false);
  const [hasMouseOver, setHasMouseOver] = useState(false);

  const inputRef = useRef(null);

  const onFocus = useCallback(() => setHasFocus(true), []);

  const onBlur = useCallback(() => setHasFocus(false), []);

  const assignInput = useCallback((input) => {
    inputRef.current = input.current;
  }, []);

  // use mouse down rather than click to accommodate click and drag events too
  const onMouseDown = useCallback(() => {
    // use a zero timeout to ensure focus is applied even on click and drag events
    setTimeout(() => inputRef && inputRef.current && inputRef.current.focus());
  }, []);

  const onMouseEnter = useCallback(() => setHasMouseOver(true), []);

  const onMouseLeave = useCallback(() => setHasMouseOver(false), []);

  const contextValue = useMemo(
    () => ({
      hasFocus,
      hasMouseOver,
      onFocus: blockGroupBehaviour ? undefined : onFocus,
      onBlur: blockGroupBehaviour ? undefined : onBlur,
      onMouseDown,
      onMouseEnter: blockGroupBehaviour ? undefined : onMouseEnter,
      onMouseLeave: blockGroupBehaviour ? undefined : onMouseLeave,
      inputRef: assignInput,
    }),
    [
      hasFocus,
      hasMouseOver,
      onFocus,
      onBlur,
      onMouseDown,
      blockGroupBehaviour,
      onMouseEnter,
      onMouseLeave,
      assignInput,
    ]
  );

  return contextValue;
};

export default useInputBehaviour;
