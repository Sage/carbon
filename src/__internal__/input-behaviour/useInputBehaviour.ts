import { useState, useRef, useCallback, useMemo } from "react";

export interface InputContextProps {
  hasFocus?: boolean;
  hasMouseOver?: boolean;
  onFocus?: () => void;
  onBlur?: () => void;
  onMouseDown?: () => void;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  inputRef?: (input: {
    current: HTMLInputElement | HTMLTextAreaElement | null;
  }) => void;
}

const useInputBehaviour = (
  blockGroupBehaviour?: boolean
): InputContextProps => {
  const [hasFocus, setHasFocus] = useState(false);
  const [hasMouseOver, setHasMouseOver] = useState(false);

  const inputRef = useRef<HTMLInputElement | HTMLTextAreaElement | null>(null);

  const onFocus = useCallback(() => setHasFocus(true), []);

  const onBlur = useCallback(() => setHasFocus(false), []);

  const assignInput = useCallback(
    (input: { current: HTMLInputElement | HTMLTextAreaElement | null }) => {
      inputRef.current = input.current;
    },
    []
  );

  // use mouse down rather than click to accommodate click and drag events too
  const onMouseDown = useCallback(() => {
    requestAnimationFrame(() => {
      inputRef?.current?.focus({
        preventScroll: true,
      });
    });
  }, []);

  const onMouseEnter = useCallback(() => setHasMouseOver(true), []);

  const onMouseLeave = useCallback(() => setHasMouseOver(false), []);

  const contextValue = useMemo(
    () => ({
      hasFocus,
      hasMouseOver,
      onFocus: blockGroupBehaviour ? undefined : onFocus,
      onBlur: blockGroupBehaviour ? undefined : onBlur,
      onMouseDown,
      onMouseEnter: blockGroupBehaviour ? undefined : onMouseEnter,
      onMouseLeave: blockGroupBehaviour ? undefined : onMouseLeave,
      inputRef: assignInput,
    }),
    [
      hasFocus,
      hasMouseOver,
      onFocus,
      onBlur,
      onMouseDown,
      blockGroupBehaviour,
      onMouseEnter,
      onMouseLeave,
      assignInput,
    ]
  );

  return contextValue;
};

export default useInputBehaviour;
