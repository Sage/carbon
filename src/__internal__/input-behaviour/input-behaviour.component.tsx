import React from "react";
import useInputBehaviour, { InputContextProps } from "./useInputBehaviour";

const InputContext = React.createContext<InputContextProps>({});

export interface InputBehaviourProps {
  children?: React.ReactNode;
}

const InputBehaviour = ({ children }: InputBehaviourProps) => {
  const contextValue = useInputBehaviour();

  return (
    <InputContext.Provider value={{ ...contextValue }}>
      {children}
    </InputContext.Provider>
  );
};

export { InputContext, InputBehaviour };
