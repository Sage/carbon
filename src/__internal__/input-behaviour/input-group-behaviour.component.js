import React from "react";
import PropTypes from "prop-types";

import useInputBehaviour from "./useInputBehaviour";

const defaultValue = {};
const InputGroupContext = React.createContext(defaultValue);

const InputGroupBehaviour = ({ children, blockGroupBehaviour }) => {
  const contextValue = useInputBehaviour(blockGroupBehaviour);

  return (
    <InputGroupContext.Provider value={contextValue}>
      {children}
    </InputGroupContext.Provider>
  );
};

InputGroupBehaviour.propTypes = {
  children: PropTypes.node,
  blockGroupBehaviour: PropTypes.bool,
};

export { InputGroupContext, InputGroupBehaviour };
