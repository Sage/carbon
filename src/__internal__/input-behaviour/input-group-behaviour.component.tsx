import React from "react";

import useInputBehaviour, { InputContextProps } from "./useInputBehaviour";
import { InputBehaviourProps } from "./input-behaviour.component";

const InputGroupContext = React.createContext<InputContextProps>({});

export interface InputGroupBehaviourProps extends InputBehaviourProps {
  blockGroupBehaviour?: boolean;
}

const InputGroupBehaviour = ({
  children,
  blockGroupBehaviour,
}: InputGroupBehaviourProps) => {
  const contextValue = useInputBehaviour(blockGroupBehaviour);

  return (
    <InputGroupContext.Provider value={contextValue}>
      {children}
    </InputGroupContext.Provider>
  );
};

export { InputGroupContext, InputGroupBehaviour };
