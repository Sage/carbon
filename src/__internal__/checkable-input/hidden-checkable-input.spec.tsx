import React from "react";
import { mount, shallow, ReactWrapper } from "enzyme";
import { InputContext, InputGroupContext } from "../input-behaviour";

import HiddenCheckableInput from "./hidden-checkable-input.component";
import HiddenCheckableInputStyle from "./hidden-checkable-input.style";

const render = (
  props = {},
  inputGroupContextValue = {},
  inputContextValue = {}
) => {
  return mount(
    <InputGroupContext.Provider value={inputGroupContextValue}>
      <InputContext.Provider value={inputContextValue}>
        <HiddenCheckableInput type="test" {...props} />
      </InputContext.Provider>
    </InputGroupContext.Provider>
  );
};

describe("HiddenCheckableInput", () => {
  let propOnBlur: jest.Mock;
  let contextOnBlur: jest.Mock;
  let groupContextOnBlur: jest.Mock;

  let propOnFocus: jest.Mock;
  let contextOnFocus: jest.Mock;
  let groupContextOnFocus: jest.Mock;

  let propOnMouseEnter: jest.Mock;
  let contextOnMouseEnter: jest.Mock;
  let groupContextOnMouseEnter: jest.Mock;

  let propOnMouseLeave: jest.Mock;
  let contextOnMouseLeave: jest.Mock;
  let groupContextOnMouseLeave: jest.Mock;

  let wrapper: ReactWrapper;

  beforeEach(() => {
    propOnBlur = jest.fn();
    contextOnBlur = jest.fn();
    groupContextOnBlur = jest.fn();

    propOnFocus = jest.fn();
    contextOnFocus = jest.fn();
    groupContextOnFocus = jest.fn();

    propOnMouseEnter = jest.fn();
    contextOnMouseEnter = jest.fn();
    groupContextOnMouseEnter = jest.fn();

    propOnMouseLeave = jest.fn();
    contextOnMouseLeave = jest.fn();
    groupContextOnMouseLeave = jest.fn();

    wrapper = render(
      {
        onBlur: propOnBlur,
        onFocus: propOnFocus,
        onMouseEnter: propOnMouseEnter,
        onMouseLeave: propOnMouseLeave,
      },
      {
        onBlur: groupContextOnBlur,
        onFocus: groupContextOnFocus,
        onMouseEnter: groupContextOnMouseEnter,
        onMouseLeave: groupContextOnMouseLeave,
      },
      {
        onBlur: contextOnBlur,
        onFocus: contextOnFocus,
        onMouseEnter: contextOnMouseEnter,
        onMouseLeave: contextOnMouseLeave,
      }
    );
  });

  it("renders as expected and sets the input role to value of the type prop by default", () => {
    const props = {
      checked: true,
      helpId: "test-help",
      name: "test-name",
      type: "test-type",
      value: "test-value",
      tabIndex: 0,
    };

    const shallowWrapper = shallow(<HiddenCheckableInput {...props} />);
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("sets the role if the prop is passed a value", () => {
    const props = {
      checked: true,
      helpId: "test-help",
      name: "test-name",
      type: "test-type",
      role: "test-role",
      value: "test-value",
      tabIndex: 0,
    };

    const shallowWrapper = shallow(<HiddenCheckableInput {...props} />);
    expect(shallowWrapper).toMatchSnapshot();
  });

  it("triggers onFocus callbacks passed from props and context", () => {
    wrapper.find(HiddenCheckableInputStyle).props().onFocus();
    expect(propOnFocus).toHaveBeenCalled();
    expect(contextOnFocus).toHaveBeenCalled();
    expect(groupContextOnFocus).toHaveBeenCalled();
  });

  it("triggers onBlur callbacks passed from props and context", () => {
    wrapper.find(HiddenCheckableInputStyle).props().onBlur();
    expect(propOnBlur).toHaveBeenCalled();
    expect(contextOnBlur).toHaveBeenCalled();
    expect(groupContextOnBlur).toHaveBeenCalled();
  });

  it("triggers onMouseEnter callbacks passed from context", () => {
    wrapper.find(HiddenCheckableInputStyle).props().onMouseEnter();
    expect(propOnMouseEnter).toHaveBeenCalled();
    expect(contextOnMouseEnter).toHaveBeenCalled();
    expect(groupContextOnMouseEnter).toHaveBeenCalled();
  });

  it("triggers onMouseLeave callbacks passed from context", () => {
    wrapper.find(HiddenCheckableInputStyle).props().onMouseLeave();
    expect(propOnMouseLeave).toHaveBeenCalled();
    expect(contextOnMouseLeave).toHaveBeenCalled();
    expect(groupContextOnMouseLeave).toHaveBeenCalled();
  });

  it("does nothing if onFocus, onBlur, onMouseEnter or onMouseLeave callbacks are not provided", () => {
    wrapper = render();
    const inputProps = wrapper.find(HiddenCheckableInputStyle).props();
    inputProps.onFocus();
    inputProps.onBlur();
    inputProps.onMouseEnter();
    inputProps.onMouseLeave();
  });

  describe("aria", () => {
    it("passes aria props to the input element", () => {
      const ariaDescribedBy = "test-aria-described-id";
      const ariaLabelledBy = "test-aria-labelled-id";
      const ariaInvalid = true;

      wrapper = render({
        "aria-describedby": ariaDescribedBy,
        "aria-labelledby": ariaLabelledBy,
        "aria-invalid": ariaInvalid,
      });
      const inputProps = wrapper.find(HiddenCheckableInputStyle).props();
      expect(inputProps["aria-describedby"]).toBe(ariaDescribedBy);
      expect(inputProps["aria-labelledby"]).toBe(ariaLabelledBy);
      expect(inputProps["aria-invalid"]).toBe(ariaInvalid);
    });
  });
});
