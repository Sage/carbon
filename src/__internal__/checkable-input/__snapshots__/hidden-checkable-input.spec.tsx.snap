// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HiddenCheckableInput renders as expected and sets the input role to value of the type prop by default 1`] = `
<styled.input
  aria-checked={true}
  checked={true}
  helpId="test-help"
  name="test-name"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="test-type"
  tabIndex={0}
  type="test-type"
  value="test-value"
/>
`;

exports[`HiddenCheckableInput sets the role if the prop is passed a value 1`] = `
<styled.input
  aria-checked={true}
  checked={true}
  helpId="test-help"
  name="test-name"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="test-role"
  tabIndex={0}
  type="test-type"
  value="test-value"
/>
`;
